src/espnow.rs

```diff
@@ -0,0 +1,183 @@
+use log::info;
+
+use esp_idf_hal::mutex::Mutex;
+use esp_idf_sys::*;
+
+static RECV_CALLBACK: Mutex<Option<Box<dyn FnMut(&[u8], &[u8]) + Send>>> = Mutex::new(None);
+static SEND_CALLBACK: Mutex<Option<Box<dyn FnMut(&[u8], SendStatus) + Send>>> = Mutex::new(None);
+
+pub static BROADCAST: [u8; 6] = [0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF];
+
+#[derive(Debug)]
+pub enum SendStatus {
+    SUCCESS = 0,
+    FAIL,
+}
+
+impl From<u32> for SendStatus {
+    fn from(val: u32) -> Self {
+        match val {
+            0 => SendStatus::SUCCESS,
+            1 => SendStatus::FAIL,
+            _ => panic!("Wrong status code"),
+        }
+    }
+}
+
+pub type PeerInfo = esp_now_peer_info_t;
+
+#[derive(Debug)]
+struct PrivateData;
+
+#[derive(Debug)]
+pub struct EspNowClient(PrivateData);
+
+impl EspNowClient {
+    pub fn new() -> Result<Self, EspError> {
```

Comment:
Since the ESP NOW client seems to be a singleton, you need to make sure that we are following the singleton pattern at least at runtime. (We don't have a way to statically enforce that (yet) in `esp-idf-svc`. We do - in `esp-idf-hal` with its "peripherals" thing.)

In other words, the first `EspNowClient::new` should succeed, and the second should fail. Please check [here](https://github.com/esp-rs/esp-idf-svc/blob/master/src/wifi.rs#L251) how this is enforced.

We also probably need to enforce that the `EspNowClient` is only created *after* the Wifi connection is initiated, by somehow passing a reference to the `EspWifi` instance into `EspNowClient`, however we can introduce this with a subsequent commit.

Reply:
Added a mutex to force singleton 👍